pkg/starknet/state.go

```diff
@@ -0,0 +1,635 @@
+
+	upd := stateUpdateResponseToStateDiff(*update)
+
+	s.updateAndCommitState(&upd, update.NewRoot, blockIterator)
+
+	// Update services
+	go s.updateServices(upd, update.BlockHash, strconv.FormatUint(blockIterator, 10))
+
+	return blockIterator + 1, update.BlockHash
+}
+
+// processPagesHashes takes an arrays of arrays of pages' hashes and
+// converts them into memory pages by querying an ethereum client.
+// notest
+func (s *Synchronizer) processPagesHashes(pagesHashes [][32]byte, memoryContract ethAbi.ABI) [][]*big.Int {
+	pages := make([][]*big.Int, 0)
+	for _, v := range pagesHashes {
+		// Get transactionsHash based on the memory page
+		hash := common.BytesToHash(v[:])
+		transactionHash, err := s.memoryPageHash.Get(hash.Hex(), starknetTypes.TransactionHash{})
+		if err != nil {
+			return nil
+		}
+		txHash := transactionHash.(starknetTypes.TransactionHash).Hash
+		log.Default.With("Hash", txHash.Hex()).Info("Getting transaction...")
+		txn, _, err := s.ethereumClient.TransactionByHash(context.Background(), txHash)
+		if err != nil {
+			log.Default.With("Error", err, "Transaction Hash", v).
+				Error("Couldn't retrieve transactions")
+			return nil
+		}
+
+		// Parse Ethereum transaction calldata for Starknet transaction information
+		data := txn.Data()[4:] // Remove the method signature hash
+		inputs := make(map[string]interface{})
+		err = memoryContract.Methods["registerContinuousMemoryPage"].Inputs.UnpackIntoMap(inputs, data)
+		if err != nil {
+			log.Default.With("Error", err).Info("Couldn't unpack into map")
+			return nil
+		}
+		// Append calldata to pages
+		pages = append(pages, inputs["values"].([]*big.Int))
+	}
+	return pages
+}
+
+// notest
+func (s *Synchronizer) updateServices(update starknetTypes.StateDiff, blockHash, blockNumber string) {
+	s.updateAbiAndCode(update, blockHash, blockNumber)
+	s.updateBlocksAndTransactions(blockHash, blockNumber)
+}
+
+// notest
+func (s *Synchronizer) updateAbiAndCode(update starknetTypes.StateDiff, blockHash string, sequenceNumber string) {
+	for _, v := range update.DeployedContracts {
+		code, err := s.feederGatewayClient.GetCode(v.Address, blockHash, sequenceNumber)
+		if err != nil {
+			return
+		}
+		log.Default.
+			With("ContractInfo Address", v.Address, "Block Hash", blockHash, "Block Number", sequenceNumber).
+			Info("Fetched code and ABI")
+		// TODO: Convert ABI and Code in Database
+		// Save the ABI
+		services.AbiService.StoreAbi(remove0x(v.Address), toDbAbi(code.Abi))
+		// Save the contract code
+		services.StateService.StoreCode(common.Hex2Bytes(v.Address), byteCodeToStateCode(code.Bytecode))
+	}
+}
+
+// notest
+func (s *Synchronizer) updateBlocksAndTransactions(blockHash, blockNumber string) {
+	block, err := s.feederGatewayClient.GetBlock(blockHash, blockNumber)
+	if err != nil {
+		return
+	}
+	log.Default.With("Block Hash", blockHash).
+		Info("Got block")
+	services.BlockService.StoreBlock(common.Hex2Bytes(blockHash), feederBlockToDBBlock(block))
```

Comment:
This is probably a bug. When syncing against L1, `updateBlocksAndTransactions` only receives the block number (see the call to `updateServices` [on line 272](https://github.com/NethermindEth/juno/blob/abba93b4ae44e02ad86fc4174d8e082c0d6a43b1/pkg/starknet/state.go#L272)). Because the hash is always empty, we'll end up overwriting the same database entry for every block.

For now, (I think) we could just pretend that the `blockNumber` is the hash and drop the `blockHash` for now, but that's definitely not a great long-term solution.

Reply:
Solved, but at that point, we don't have a blockHash if we are executing against L1, so I wait until fetched the block and later I use the same blockHash from the fetched block to store it